### ISSUE TITLE
Regarding the loss function calculation in the code

### ISSUE THREAD

**Issue by dr-smgad** [princeton-nlp/SimCSE#47]
Hi,

Thanks for sharing this very helpful implementation.

I have a question about the a difference between the loss design in the paper and in the code as in the code I could find it as :

```labels = torch.arange(cos_sim.size(0)).long().to(cls.device)
    loss_fct = nn.CrossEntropyLoss()

    # Calculate loss with hard negatives
    if num_sent == 3:
        # Note that weights are actually logits of weights
        z3_weight = cls.model_args.hard_negative_weight
        weights = torch.tensor(
            [[0.0] * (cos_sim.size(-1) - z1_z3_cos.size(-1)) + [0.0] * i + [z3_weight] + [0.0] * (z1_z3_cos.size(-1) - i - 1) for i in range(z1_z3_cos.size(-1))]
        ).to(cls.device)
        cos_sim = cos_sim + weights

    loss = loss_fct(cos_sim, labels)
```

This does not show the denominator component in the loss function shown in the paper Eq.5 (i.e., normalization with all negative samples in the batch). Also, what does the `labels` vector mean here? It is an arrange of values with start=0 and end=cos_sim.size(0) ... what does this mean when used with the `nn.CrossEntropyLoss()` loss function?

**Comment by gaotianyu1350** [comment]
Hi,

`cos_sim[x, y]` is the cosine similarity between the xth sentence and the yth sentence in the batch (in the supervised case, the xth premise and the yth hypothesis). Here in the loss function we regard the first dimension as the instance and the second dimension as the "class", so `labels` should be a list from 0 to batch_size - 1. The last line of the code you show here shows the "denominator component" in the loss because it's implemented in Softmax Cross Entropy.

**Comment by dr-smgad** [comment]
Thanks for your reply. To make sure that I understand you well, when you mention x and y in the **_unsupervised setup_** does this mean the same sentence with pooler output from two adjacent hidden layers from the encoder output as I can see in the code `z1, z2 = pooler_output[:,0], pooler_output[:,1]` and also for the hard negative it is a deeper layer `z3 = pooler_output[:, 2]`?

For the labels part, now I understand the logic behind its size as for each sample in the batch we need a label, but I can't get the idea behind its values. For example, lets say we have three samples and we measure the similarity between their two different representations in a vector `[0.5, 0.2, 0.8]` and the labels vector will be `[0, 1, 2]` now how does this is supposed to work as true labels for the cross entropy? like for the first example we get `y_pred=0.5 and y_true=0` or the second `y_pred=0.2 and y_true=1`

**Comment by gaotianyu1350** [comment]
Hi,

The second dimension of `pooler_output` does not represent different layers, but simply different sentences. 

About cross entropy, you should check [this pytorch page](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html) for its interface. In short, the "labels" are the true "classes" of each instance.

**Comment by dr-smgad** [comment]
Thanks for the clarification. I got your point for the cross entropy function. For the pooler, if this is the sentence dimension, then why we fix similarity comparisons at sentence 0, 1 , and 2  (if there is a hard negative), while the minibatch has 512 sentences as per the paper. This is my final question in this thread and I will close it upon your reply. Thanks again for your feedback.

**Comment by gaotianyu1350** [comment]
For `pooler_output`, the first dimension is the batch size and for each "instance" in the batch, there are two sentences (if using hard negatives, then there are three sentences), and they can be referred by the second dimension.

**Comment by dr-smgad** [comment]
Ok, so you feed these sentences as one input with separator token to the tokenizer, but this will limit their allowed size to be less than 512 for the two (or three combined). Thanks.
